Add async methods in System.Data.Common, implement IAsyncDisposable

roji opened the pull request:
Closes #35012

ajcvickers commented on src/System.Data.Common/src/System/Data/Common/DbCommand.netcoreapp.cs:
```diff
@@ -0,0 +1,38 @@
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+            else
```
Redundant else

roji replied:
Will clean up tomorrow, thanks.

roji replied:
Done.

roji replied:
Will do the same for the rest of the project in a separate PR.

stephentoub commented on src/System.Data.Common/src/System/Data/Common/DbCommand.netcoreapp.cs:
```diff
@@ -0,0 +1,38 @@
+        public virtual ValueTask DisposeAsync()
+        {
+            Dispose();
+            return new ValueTask(Task.CompletedTask);
```
This can just be `return default;`.

stephentoub commented on src/System.Data.Common/src/System/Data/Common/DbDataReader.netcoreapp.cs:
```diff
@@ -0,0 +1,38 @@
+        public virtual ValueTask DisposeAsync()
+        {
+            Dispose();
+            return new ValueTask(Task.CompletedTask);
```
Ditto

stephentoub commented on src/System.Data.Common/src/System/Data/Common/DbTransaction.netcoreapp.cs:
```diff
@@ -0,0 +1,58 @@
+        public virtual ValueTask DisposeAsync()
+        {
+            Dispose();
+            return new ValueTask(Task.CompletedTask);
```
Ditto

stephentoub commented on src/System.Data.Common/src/System/Data/Common/DbConnection.netcoreapp.cs:
```diff
@@ -0,0 +1,83 @@
+        public virtual ValueTask DisposeAsync()
+        {
+            Dispose();
+            return new ValueTask(Task.CompletedTask);
```
Ditto

stephentoub commented on src/System.Data.Common/ref/System.Data.Common.netcoreapp.cs:
```diff
@@ -0,0 +1,38 @@
+        public virtual System.Threading.Tasks.Task CloseAsync(System.Threading.CancellationToken cancellationToken = default) => throw null;
+        public virtual System.Threading.Tasks.ValueTask DisposeAsync() => throw null;
+    }
+}
```
Tests?

roji replied:
You're referring to tests which would verify that the async versions delegate to the sync ones, right (and that they throw on cancelled token)? None currently exist for existing async wrappers - if you think that's valuable I can do a separate PR that adds them to the entire project.

stephentoub replied:
Yup.

roji replied:
Will do.

roji commented:
ping @divega for an extra set of eyes

roji commented:
@divega you're right, I did a test build in #38051 and it works - seems like the issue was with the tests in #36123, and we currently don't have any in this PR. Rebased, squashed and removed the netcoreapp special-casing.
